Added Questions - Shooter

AustinFortunato opened the pull request:
BAM!

AustinFortunato committed:
Added Questions - Shooter

AustinFortunato committed:
Added Questions - Shooter

AustinFortunato committed:
Answered the Questions

In Grave Detail

AustinFortunato committed:
Answered the Questions

In Grave Detail

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -14,4 +28,5 @@ public Shooter(RobotMap.ShooterMap map) {
         shooterWheelMotor = map.shooterWheel();
     }
 
+    public 
```
This is gonna be fun...

AustinFortunato committed:
Added the spin motor command

spinMotor() takes 1 parameter, speed!

AustinFortunato committed:
Added the spin motor command

spinMotor() takes 1 parameter, speed!

AustinFortunato committed:
Motor commands and maths

Change motor speed and stop it, and then! MATH! calculateVelocity gets the velocity needed to hit a target with a given angle of launch.

AustinFortunato committed:
Motor commands and maths

Change motor speed and stop it, and then! MATH! calculateVelocity gets the velocity needed to hit a target with a given angle of launch.

AustinFortunato committed:
Deleted garbage

IDK what i'm doing but that ARG definitely wasn't supposed to be there...

AustinFortunato committed:
Deleted garbage

IDK what i'm doing but that ARG definitely wasn't supposed to be there...

AustinFortunato committed:
Checks that the target is reachable

IDK what it should return if it isn't so it just returns 0 right now.

AustinFortunato committed:
Checks that the target is reachable

IDK what it should return if it isn't so it just returns 0 right now.

AustinFortunato committed:
Merge remote-tracking branch 'origin/Shooter' into Shooter

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -1,17 +1,71 @@
 
-    private SendableSpeedController shooterWheelMotor;
+    private final SendableSpeedController shooterWheelMotor;
+    public static double speed;
```
We don't need this variable.

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -1,17 +1,71 @@
+    public static double speed;
+    public static double gravity;
+    public static double theta;
+    public static double y;
```
This should be `TARGET_HEIGHT`

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -1,17 +1,71 @@
+    private final SendableSpeedController shooterWheelMotor;
+    public static double speed;
+    public static double gravity;
+    public static double theta;
```
`gravity` and `theta` should be final and in upper snake case. Also set the values here.

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -1,17 +1,71 @@
 
+    public CommandBase spinUp() {
+        return new InstantCommand(() -> {
+            shooterWheelMotor.set(speed);
```
Eventually this should call the `calculateVelocity()` and some other function that will calculate the RPMs.

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -1,17 +1,71 @@
+     * angle (theta).
+     */
+
+    public static double calculateVelocity(double x) {
```
Please rename `x` to `distanceToTarget`.

AustinFortunato committed:
Changed variable names

Also made sure it accounted for the height of the shooter.

Please work.

AustinFortunato committed:
Changed variable names

Also made sure it accounted for the height of the shooter.

Please work.

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -20,23 +19,22 @@
     public CommandBase spinUp() {
         return new InstantCommand(() -> {
-            shooterWheelMotor.set(speed);
+            shooterWheelMotor.set(8.5); // Make a calculateRPM() function or something to get the speed
```
This should probably not be `8.5` since it's limited to `1.0`.

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -48,21 +46,21 @@ public CommandBase spinDown() {
-            double tanSide = x * Math.tan(Math.toRadians(theta)) - y;
-            double cosSide = Math.cos(Math.toRadians(theta)) * Math.cos(Math.toRadians(theta));
+        if (distanceToTarget * Math.tan(Math.toRadians(THETA)) >= verticalDistance) {
+            final double GRAVITY_SIDE = GRAVITY * distanceToTarget * distanceToTarget;
```
Sorry I led you astray here. These don't need to be in all caps. Just the `final static` ones above.

bot190 commented on src/main/java/frc/robot/maps/RobotMap.java:
```diff
@@ -32,10 +32,15 @@ public IntakeMap getIntakeMap() {
         }
+
+        public double shooterHeight() {
+            return 45;
```
This should default to 0 or not have a default at all. This value should be in "Tempest Map"

bot190 reviewed (commented):
Please provide some more details about the algorithm you implemented.

DavidDesfosse committed:
Modified shooter to use a PID speed controller

AustinFortunato committed:
Merge remote-tracking branch 'origin/Shooter' into Shooter

AustinFortunato committed:
Explanation of the velocity formula

it's like a lot of math..

AustinFortunato committed:
It works i swear

lol please

AustinFortunato committed:
Added tempest shooter height

It's very cool- i swear

AustinFortunato committed:
Gets distance now!

baby steps

AustinFortunato committed:
Beautiful

msoucy commented on src/main/java/frc/robot/maps/RobotMap.java:
```diff
@@ -44,13 +46,91 @@ public IntakeMap getIntakeMap() {
     public static class ShooterMap {
-        public SendableSpeedController shooterWheel() {
-            return new MockSpeedController();
+        public PIDSpeedController shooterWheel() {
```
We've got to stick this in chopshoplib... but for now this is fine.

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -2,16 +2,67 @@
+    private final SendableSpeedController shooterWheelMotor;
+    public final double shooterHeight;
+    public static double verticalDistance;
+    public final static double GRAVITY = 386.09;
```
What unit is gravity in?

AustinFortunato replied:
As benjamin requested, freedom units- specifically inches

msoucy replied:
That was a subtle hint. Change the variable name to `GRAVITY_IN_PER_SECOND_SQ`, or otherwise leave a comment saying what unit it's in.

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -2,16 +2,67 @@
+        }, this);
+    }
+
+    public CommandBase spinDown() {
```
This is less "spin down" and more "halt", but sure, that works

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -2,16 +2,67 @@
+
+    public CommandBase spinDown() {
+        return new InstantCommand(() -> {
+            shooterWheelMotor.stopMotor();
```
If you're feeling snazzy, you can replace this with `return new InstantCommand(shooterWheelMotor::stopMotor, this);`, it's just a little more terse (but not required)

msoucy commented on src/main/java/frc/robot/subsystems/Calculate-trajectory.md:
```diff
@@ -0,0 +1,53 @@
+# Simplifying trajectory math
```
I turned this into a wiki page - remove the file from the branch please

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -2,16 +2,67 @@
+        if (distanceToTarget * Math.tan(Math.toRadians(THETA)) >= verticalDistance) {
+            final double gravitySide = GRAVITY * distanceToTarget * distanceToTarget;
+            final double tanSide = distanceToTarget * Math.tan(Math.toRadians(THETA)) - verticalDistance;
+            final double cosSide = Math.cos(Math.toRadians(THETA)) * Math.cos(Math.toRadians(THETA));
```
Store `Math.cos(Math.toRadians(THETA))` in a variable. Or even better, make the definition of `THETA` be in radians to begin with.

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -2,16 +2,67 @@
+            final double cosSide = Math.cos(Math.toRadians(THETA)) * Math.cos(Math.toRadians(THETA));
+
+            final double velocity = Math.sqrt(gravitySide / tanSide / cosSide / 2);
+            return velocity;
```
Don't make a variable just to return it. Return the expression directly.

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -2,16 +2,67 @@
+     * gravity.
+     */
+
+    public static double calculateVelocity(final double distanceToTarget) {
```
Is this actually used anywhere?

AustinFortunato replied:
No it will be though- we need to figure out how to turn the shooters motors into inches/second first.

AustinFortunato committed:
Did what Matt wanted

Full filled the beast's requests.

AustinFortunato committed:
I'm sorry ben

to be fair, Vision returns meters so like...

AustinFortunato committed:
Changed some silly comments

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -2,16 +2,66 @@
+    public final double shooterHeight;
+    public static double verticalDistance;
+    // meters/second
+    public final static double GRAVITY = 9.81;
```
Please make these freedom units again.

bot190 reviewed (changes requested):
Please add a function to convert linear velocity to RPM based on the shooter wheel size of 4" it should apply a loss of 20% (eg 80% of the calculated value). Then call that and `calculateVelocity` in `spinUp`.

AustinFortunato committed:
Freedom units, calculates RPM

bitte.

AustinFortunato committed:
Sorry oh unneeded commits...

Forgot it loses 20% of its velocity.

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -2,16 +2,78 @@
+
+    // inches/second
+    public final static double GRAVITY = 386.2205;
+    // icnhes
```
Spelling issue here.

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -2,16 +2,78 @@
 public class Shooter extends SubsystemBase {
 
-    private SendableSpeedController shooterWheelMotor;
+    private final SendableSpeedController shooterWheelMotor;
```
This should be `PIDSpeedController`. We'll want to set the reference point instead of the normal `set` command.

bot190 reviewed (commented):
This is getting close. We should prepare for using PID to maintain velocity. Though for now we don't want to actually use the PID stuff (we need to tune the PID constants).

AustinFortunato committed:
PIDs

gets PID stuffs

AustinFortunato committed:
More PID

bot190 commented on src/main/java/frc/robot/maps/FrancoisMap.java:
```diff
@@ -60,6 +60,8 @@ public ShooterMap getShooterMap() {
             public PIDSparkMax shooterWheel() {
                 CANSparkMax leader = new CANSparkMax(25, MotorType.kBrushless);
                 CANSparkMax follower = new CANSparkMax(26, MotorType.kBrushless);
+                leader.setInverted(true);
```
I wouldn't expect both of these to be inverted. For now we should probably not invert any of these.

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -1,17 +1,79 @@
+
+    public CommandBase spinUp() {
+        return new InstantCommand(() -> {
+            shooterWheelMotor.set(calculateRPM() / 5800);
```
5800 should probably be a `final` value and also probably closer to 5000 or so.

bot190 reviewed (changes requested):
Let's get this finished and merged!

AustinFortunato committed:
Fixed what ben wanted

made max RPM a variable and uninverted the motor
